File: client/src/main/java/com/fiwio/iot/demeter/app/demeter/discovery/ResolveListener.java
package com.fiwio.iot.demeter.app.demeter.discovery;

import android.net.nsd.NsdManager;
import android.net.nsd.NsdServiceInfo;
import android.os.Build;
import android.os.Handler;
import android.support.annotation.RequiresApi;
import android.util.Log;

@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
public class ResolveListener implements NsdManager.ResolveListener {

    private static final String LOG_TAG = ResolveListener.class.getSimpleName();
    private final ClientNetworkListener clientNetworkListener;
    private final Handler handler;

    public ResolveListener(ClientNetworkListener clientNetworkListener, Handler handler) {
        this.clientNetworkListener = clientNetworkListener;
        this.handler = handler;
    }


    @Override
    public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
        Log.e(LOG_TAG, "Service resolve error (" + errorCode + ")");
        handler.post(new Runnable() {
            @Override
            public void run() {
                clientNetworkListener.onServiceDiscoveryError();
            }
        });
    }


    @Override
    public void onServiceResolved(final NsdServiceInfo serviceInfo) {
        Log.d(LOG_TAG, "Service resolve success (" + serviceInfo + ")");
        handler.post(new Runnable() {
            @Override
            public void run() {
                clientNetworkListener.onServiceDiscovered(
                        new HostInfo(
                                serviceInfo.getServiceName(),
                                serviceInfo.getHost(),
                                serviceInfo.getPort()));
            }
        });
    }
}

File: mobile-networking/src/main/java/com/fiwio/iot/demeter/android/networking/aspects/discovery/ResolveListener.kt
package com.fiwio.iot.demeter.android.networking.aspects.discovery

import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.os.Build
import android.os.Handler
import android.support.annotation.RequiresApi
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}


@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
class ResolveListener(private val handler: Handler, private val clientNetworkListener: ClientNetworkListener) : NsdManager.ResolveListener {

    override fun onResolveFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
        logger.error { "Service resolve error ($errorCode)" }
        handler.post { clientNetworkListener.onServiceDiscoveryError() }
    }

    override fun onServiceResolved(serviceInfo: NsdServiceInfo) {
        logger.debug { "Service resolve success ($serviceInfo)" }
        handler.post {
            clientNetworkListener.onServiceDiscovered(
                    HostInfo(
                            serviceInfo.serviceName,
                            serviceInfo.host,
                            serviceInfo.port))
        }
    }
}
